### DISCUSSION TITLE
Custom data in error response

### DISCUSSION THREAD

**Discussion by SerheyDolgushev** [api-platform/api-platform#2631] [Ideas]
## Problem
For errored requests, API Platform returns the string representation of the error in the `detail` field. Which works fine for most cases. However, in some cases, additional information about the error needs to be returned as a non-string structure.

### Steps to reproduce
Let's assume there is following exception in the system (it stores an array of strings in `errorCodes` property):
```php
<?php

class ExceptionWithErrorCodes extends \RuntimeException
{
    public function __construct(private array $errorCodes, protected $message) {
        parent::__construct($this->message);
    }a

    public function getErrorCodes(): array
    {
        return $this->errorCodes;
    }
}
```

In some cases one of the API Platform Processors throws this exception:
```php
...
throw new ExceptionWithErrorCodes(['CODE1', 'CODE2'], 'Human-readable message');
```

The following code will be executed once this exception is thrown:
1. `ApiPlatform\Symfony\EventListener\ErrorListener::duplicateRequest` [will return a duplicate request](https://github.com/api-platform/core/blob/86be7dffc121c29aa964ce1b56e78a0b19dfcaa0/src/Symfony/EventListener/ErrorListener.php#L97).
2. `ApiPlatform\Action\ExceptionAction::__invoke` will be used to handle step 1. And it [will convert exception to FlattenException](https://github.com/api-platform/core/blob/86be7dffc121c29aa964ce1b56e78a0b19dfcaa0/src/Action/ExceptionAction.php#L53). At this point `errorCodes` field of original `ExceptionWithErrorCodes` will be lost.
3. `ApiPlatform\Problem\Serializer\ErrorNormalizer::normalize` will be used to [convert exception from the previous step into the response](https://github.com/api-platform/core/blob/86be7dffc121c29aa964ce1b56e78a0b19dfcaa0/src/Problem/Serializer/ErrorNormalizer.php#L48-L61).
4. Response will return JSON array. It's detail field will be `Human-readable message`. But `['CODE1', 'CODE2']` of the original exception will be not reflected in the response.

#### Solution 1

1. Override `ApiPlatform\Action\ExceptionAction` with the following changes:
    1. Do not convert all exceptions into FlattenException, and use Throwable instead:
        `public function __invoke(FlattenException $exception, Request $request): Response`
        =>
        `public function __invoke(\Throwable $exception, Request $request): Response`
    2. Extract `errorCodes` from `$exception` and pass it as part of `$context` into serializer
        ```php        
        ...
        if ($exception instanceof ExceptionWithErrorCodes) {
            $context['errorCodes'] = $exception->getErrorCodes();
        }
        ...
        ```
2. Implement a new serializer, that will read `errorCodes` from the `$context` and inject it into the response:
    ```php
        $data = [
            'type' => $context[self::TYPE] ?? $this->defaultContext[self::TYPE],
            'title' => $context[self::TITLE] ?? $this->defaultContext[self::TITLE],
            'error_codes' => $context['errorCodes'] ?? [],
        ];

        if ($this->debug) {
            $data['trace'] = $object->getTrace();
        }

        return $data;
    ```
#### Solution 2

1. Modify `ExceptionWithErrorCodes` and store the serialized error codes in its message (as string).
2. Implement a new serializer that will check if the exception is `ExceptionWithErrorCodes`, and if so it will return error codes from it's unserialized message.

## Goal of this discussion
**Solution 1** seems to be the less hacky one. But it requires overriding `ApiPlatform\Action\ExceptionAction` and using internal traits/classes. The main goal of this discussion is to start a conversation and discuss if that is a good idea to modify `ApiPlatform\Action\ExceptionAction` so it can be easily extended (without using internal traits/classes) in cases like this one.

**Comment by soyuka** [comment]
Hello, the detail part of the [specification](https://datatracker.ietf.org/doc/html/rfc7807#section-3.1) has: 

> "detail" (string) - A human-readable explanation specific to this occurrence of the problem.

Therefore you should not use `detail` for this but another error field. Let's assume we want to add `errorCodes` there are two recommended ways in API Platform 3.2: 

- declare an error resource with your additional field https://api-platform.com/docs/guides/error-resource/
- create an error provider https://api-platform.com/docs/guides/error-provider/ 

I'd declare an [error resource](https://api-platform.com/docs/guides/error-resource/) as it's cleaner to have a domain exception that is also an Error api-wise.

**Reply by SerheyDolgushev** [reply]
@soyuka thank you very much for your reply

I checked the [error resource](https://api-platform.com/docs/guides/error-resource/) resource example you provided, but it does not solve the original issue: **getting additional exception data in the API error response.**

Let's assume we added a new `array<String> error_codes` property into `MyDomainException`. How can we include that in the API Response? Right now, seems like there is no simple way to do that. The only possible way is to override ApiPlatform\Action\ExceptionAction which uses internal traits and depends on deprecated classes.

The main reason why I opened this discussion is to find out if that is a common scenario, and if it is worth making it easily achievable.

**Reply by soyuka** [reply]
I don't understand, when the `MyDomainException` is a **Resource**, all its field will be serialized you don't have more to do to get additional data in your response.

**Reply by SerheyDolgushev** [reply]
I just tested, and indeed after the new property is added to the exception class, it is included in the API response. @soyuka thank you very much! :)
